MsApplicationPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -48,6 +97,210 @@ StatusToHexString (
   return StatusString;
 }
 
+/**
+ * Prints an error message to the console
+ * @param[in] Message     The message to print
+ * @param[in] Status      The EFI_STATUS to print
+*/
+VOID
+PrintError (
+  CHAR16      *Message,
+  EFI_STATUS  Status
+  )
+{
+  mST->ConOut->OutputString (mST->ConOut, Message);
+  mST->ConOut->OutputString (mST->ConOut, L"Error: 0x");
+  mST->ConOut->OutputString (mST->ConOut, StatusToHexString (Status));
+  mST->ConOut->OutputString (mST->ConOut, L"\r\n");
+}
+
+/**
+  This function checks if a specified certificate is present in the EFI image security database (db).
+
+  @param  Certificate      A pointer to the CERTIFICATE_ENTRY structure that contains the certificate to search for.
+  @param  OptionalStatus   If not NULL, a pointer to a variable that receives the status of the operation.
+
+  @retval TRUE             The certificate was found in the database.
+  @retval FALSE            The certificate was not found in the database.
+                           If OptionalStatus is not NULL, it contains an error code indicating the reason for the failure.
+
+**/
+BOOLEAN
+IsCertificateInDB (
+  IN  CERTIFICATE_ENTRY  *Certificate,
+  OUT EFI_STATUS         *OptionalStatus
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      Attributes;
+  UINTN       DbSize;
+  UINTN       Index;
+  UINTN       CertSize;
+  UINT8       *Cert;
+  UINTN       CertCount;
+  UINT8       *Db;
+  BOOLEAN     Found;
+
+  EFI_SIGNATURE_LIST  *CertList;
+  EFI_SIGNATURE_DATA  *CertData;
+
+  Found      = FALSE;
+  Db         = NULL;
+  DbSize     = 0;
+  Attributes = VARIABLE_ATTRIBUTE_NV_BS_RT_AT | EFI_VARIABLE_APPEND_WRITE;
+
+  if ((mST == NULL) || (Certificate == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  //
+  // Read the size and attributes of the DB
+  //
+  Status = mRT->GetVariable (
+                  L"db",
+                  &gEfiImageSecurityDatabaseGuid,
+                  &Attributes,
+                  &DbSize,
+                  NULL
+                  );
+
+  //
+  // Confirm we got the expected error that the buffer was too small
+  //
+  if (Status != EFI_BUFFER_TOO_SMALL) {
+    //
+    // Likely this will continue to fail on reboot
+    //
+    return FALSE;

```

Comment:
Should EFI_BUFFER_TOO_SMALL get bubbled up through OptionalStatus?

Reply:
This is actually the expected error, since I'm giving it a NULL buffer with a size of 0. It in return tells me the full size that I can then use to allocate. Are you asking if I should replace EFI_BUFFER_TOO_SMALL with EFI_SUCCESS after this check?